test/test_framework.py

```diff
@@ -43,17 +44,18 @@ def setUp(self):
     def test_deprecated_init(self):
         # For 0.7, this still works, but it is deprecated.
         with self.assertLogs(level="WARNING") as cm:
-            framework = ops.Framework(':memory:', None, None, None)
+            framework = ops.Framework(':memory:', None, None, None)  # type: ignore
         self.assertIn(
             "WARNING:ops.framework:deprecated: Framework now takes a Storage not a path",
             cm.output)
         self.assertIsInstance(framework._storage, SQLiteStorage)
 
     def test_handle_path(self):
         cases = [
-            (ops.Handle(None, "root", None), "root"),
+            (ops.Handle(None, "root", 'test1'), "root[test1]"),
```

Comment:
Unless I'm missing something, I'm a bit nervous that we're changing / removing these test cases. If anything I think we should keep the existing ones (changing the typing shouldn't change them) and add a couple of new ones testing the additional cases.

Reply:
The `Handle.__init__` code is like this:

```python
    def __init__(self, parent: Optional[Union['Handle', 'Object']], kind: str, key: str):
        ...
        self._key = key
        ...
            if key:
                self._path = f"{parent}/{kind}[{key}]"
            else:
                self._path = f"{parent}/{kind}"
```

The docs say (my emphasis):

> The handle key is a **string** uniquely identifying the object.

The key is also used for hashing and equality checks (`None` is fine in either of those situations, but so are other non-string things). It's also a public member, and I don't have the experience to know how that's generally used, however, it does claim to be a string as well:

```python
    @property
    def key(self) -> str:
```

This seems messy to me. Shouldn't we either have the type signature and documentation match the behaviour and accept `str|None` (the docs could say that this *should not* be done), or have the implementation match the type signature and documentation and reject `None` as an invalid value for `key`?

Reply:
Hmm, yeah. From looking at the code, I think `key` should have been an `Optional[str]` all along -- the fact that is was typed as `str` was I think an oversight. The `if key` would handle empty string as well, but I think `None` would be the intention.

I've found one charm that passes `None` as the `key` argument to `Handle`.

So I think we should change `Handle`'s key arg to `Optional[str]`, and same for `nest`'s key argument. Those should be perfectly backwards-compatible changes, as we're expanding the type.

I also think we should change the `key` property's type to be `Optional[str]`. That isn't quite a backward-compatible change (typing-wise), because code that accesses the property will then have to `assert key is not None` to type check. But I've done a quick search and I can't see any charms using that property (it's a bit advanced/arcane to use `Handle` directly in the first place, let alone access `handle.key`).

So I think it's safe to change and we should do it. If you want, we could start a file `CHANGES.txt` that we use to record notes and breaking changes for the upcoming release, so we remember.

Reply:
Done - lmk if you'd prefer a particular format for the changes file.